pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go

```diff
@@ -177,18 +177,39 @@ func (c *clusterMemberRemovalController) attemptToScaleDown(ctx context.Context,
 	if err != nil {
 		return fmt.Errorf("could not get a list of unhealthy members: %v", err)
 	}
+
+	// map machines pending deletion to node internal ip
+	votingMachinesPendingDeletionIndex := ceohelpers.IndexMachinesByNodeInternalIP(votingMachinesPendingDeletion)
+	// find unhealthy machine pending deletion to remove first from the cluster, without violating quorum
+	var unhealthyMachinePendingDeletion []*machinev1beta1.Machine
 	if len(unhealthyMembers) > 0 {
 		var unhealthyMembersURLs []string
 		for _, unhealthyMember := range unhealthyMembers {
+			unhealthyMemberNodeInternalIP, err := ceohelpers.MemberToNodeInternalIP(unhealthyMember)
+			if err != nil {
+				return fmt.Errorf("cannot get node internal ip for unhealthy member %v: %w", unhealthyMember, err)
+			}
+			unhealthyMemberMachine, ok := votingMachinesPendingDeletionIndex[unhealthyMemberNodeInternalIP]
+			if ok {
+				unhealthyMachinePendingDeletion = append(unhealthyMachinePendingDeletion, unhealthyMemberMachine)
+			}
 			if len(unhealthyMember.PeerURLs) > 0 {
 				unhealthyMembersURLs = append(unhealthyMembersURLs, unhealthyMember.PeerURLs[0])
 			} else {
 				unhealthyMembersURLs = append(unhealthyMembersURLs, unhealthyMember.Name)
 			}
 		}
-		return fmt.Errorf("cannot proceed wth scaling down, unhealthy members found: %v", unhealthyMembersURLs)
+		if len(unhealthyMachinePendingDeletion) > 0 {
+			klog.V(4).Infof("found unhealthy voting members with machine pending deletion: %v", unhealthyMachinePendingDeletion)
+			klog.V(4).Infof("unhealthy members found: %v", unhealthyMembersURLs)
+		} else {
+			return fmt.Errorf("cannot proceed wth scaling down, unhealthy members found: %v", unhealthyMembersURLs)
```

Comment:
maybe it's already too late for my brain, but wouldn't this never delete a node when there are only healthy members?

Reply:
okay wait, this is confusing:

> if len(unhealthyMembers) > 0 

so there are some unhealthy members, then we join with the pending deletions. If there are no unhealthy members pending deletion we are returning an error?

gotta look at this tomorrow again :) let me know what you think @Elbehery

Reply:
Gr8 Question .. 

So if this  `if len(unhealthyMembers) > 0` condition is true, then we have unhealthy members. Usually we should **not** allow scaling down in this case, however, if the unhealthy voting member is `pending deletion` && N+1 voting members exist, where N is the `desiredControlPlaneReplicaCount` .. then we should remove the unhealthy voting member whose machine is pending deletion .. 

On the other hand, when the condition `if len(unhealthyMembers) > 0` is false, then we skip this whole block and proceed to delete voting members who machine is pending deletion .. 

cc @hasbro17 @JoelSpeed

Reply:
How would we get to `N+1` voting members? Normally we have `N`, one goes unhealthy, we can't add a new member while one of the `N` is unhealthy, we can only remove. So as far as i can tell, as soon as one voting member goes unhealthy, the only way we can go is down from `N`.

Do we mean pods here rather than voting members?

Reply:
> cannot proceed with scaling down, unhealthy members found

The voting members are etcd pods which are members of the etcd cluster and **not** learner

> How would we get to N+1 voting members

Assuming N which is `desiredControlPlaneReplicaCount` is `3` 

We added a new master machine, so now we have `4` master machine (i.e. voting etcd members). Now one goes unhealthy / failed. Running `oc delete machine` on the failed machine, will remove the corresponding etcd voting member from the etcd cluster. This is allowed since we have N+1 members while the required for Quorum safety is only N

> Normally we have N, one goes unhealthy, we can't add a new member while one of the N is unhealthy, we can only remove . So as far as i can tell, as soon as one voting member goes unhealthy, the only way we can go is down from N.

So as discussed with @hasbro17 , currently we do not allow scaling up while the cluster is unhealthy.

Reply:
> We added a new master machine, so now we have 4 master machine (i.e. voting etcd members). Now one goes unhealthy / failed.

The scenario I would like to resolve is what happens when a Machine goes unhealthy *before* you are able to add the 4th master machine. This is likely what will happen in the real world. A machine goes down for some reason, so you want to replace that, you ask for the machine to be deleted, you expect everything to recover, right now it won't, and if we rely on this N+1 voting members instead of just N+1 pods, I still don't believe that will resolve the issue.

When the cluster is unhealthy and there is a new node waiting to be added as a learner, IMO, we should scale down to allow the new learner to be added. I don't think what's described is resolving that issue just yet

Reply:
> When the cluster is unhealthy and there is a new node waiting to be added as a learner, IMO, we should scale down to allow the new learner to be added. I don't think what's described is resolving that issue just yet 

this most probably violates Quorum safety, and needs a new design

Reply:
wdyt herr @tjungblu ?

Reply:
> The scenario I would like to resolve is what happens when a Machine goes unhealthy before you are able to add the 4th master machine. This is likely what will happen in the real world. A machine goes down for some reason, so you want to replace that,

Yeah that's true, it's unsafe however. On n=3, one unhealthy node already brings you to the point where you have to still have quorum but any additional failure will break quorum. Adding a learner in that scenario will make it more likely that you overload the remaining two that also must handle both reads + writes of the usual cluster traffic. Imagine that third member was overloaded and thus became unhealthy - now you're in a cascading failure.

It's still better IMHO to have a human go and evaluate the health of the lost member and ensure that the others are sufficiently able to handle another learner member or restore directly from snapshot. Having said that, this is just being conservative on my end. Maybe @hasbro17 has another take on this.

Since https://github.com/openshift/cluster-etcd-operator/pull/872 we're also stopping to rollout any static revision as this will break quorum. Before the PR adding a new learner would definitely create a new static pod revision and that certainly broke quorum in lots of clusters already rendering them unusable.

Reply:
> It's still better IMHO to have a human go and evaluate the health of the lost member and ensure that the others are sufficiently able to handle another learner member or restore directly from snapshot

I guess my retort to that is that, if the user deleted the Machine, that's their signal this is going away, but maybe you think the second, oh right, I need to check etcd, step is of additional help

Reply:
So just to clarify, this bug/PR is specifically meant to address the scenario of removing an unhealthy member when we have >= N healthy voting members present in addition, e.g deleting the machine for the unhealthy member in a cluster with 3/4 healthy members.

@JoelSpeed I agree that the real world scenario is more likely to be wanting to delete the 1/N unhealthy machines and expecting to auto-recover. As you've mentioned, that's not possible today (since that was a non-goal of the quorum protection enhancement) but we can certainly try and address it now.

I've created a ticket for us to track that work separately given it's not going to be a simple change without considering the effects of the quorum checker that we have to prevent revision rollouts that would cause quorum loss. So let's follow up the discussion on there.
https://issues.redhat.com/browse/ETCD-328